pkg/workspace/resource/node_claim.go

```diff
@@ -48,179 +49,203 @@ func NewNodeClaimManager(c client.Client, recorder record.EventRecorder, expecta
-			klog.ErrorS(updateErr, "failed to update NodeClaim status condition", "workspace", klog.KObj(wObj))
-		}
+// ScaleUpNodeClaims scales up the NodeClaims for the given workspace
+// this function will be invoked before creating workloads for workspace in order to ensure nodes.
+func (c *NodeClaimManager) ScaleUpNodeClaims(ctx context.Context, wObj *kaitov1beta1.Workspace, nodesToCreate int) (bool, error) {
+	workspaceKey := client.ObjectKeyFromObject(wObj).String()
+	klog.InfoS("Scaling up additional NodeClaims", "workspace", workspaceKey, "toCreate", nodesToCreate)
 
-		c.expectations.ExpectCreations(c.logger, workspaceKey, nodesToCreate)
-
-		nodeOSDiskSize := c.determineNodeOSDiskSize(wObj)
-
-		for range nodesToCreate {
-			var nodeClaim *karpenterv1.NodeClaim
-			var err error
-			created := false
-
-			retry.RetryOnConflict(retry.DefaultRetry, func() error {
-				nodeClaim = nodeclaim.GenerateNodeClaimManifest(nodeOSDiskSize, wObj)
-				err = c.Client.Create(ctx, nodeClaim)
-
-				if err == nil {
-					created = true
-					return nil
-				}
-				return err
-			})
-
-			if !created {
-				// Failed to create, decrement expectations
-				c.expectations.CreationObserved(c.logger, workspaceKey)
-				if err != nil {
-					c.recorder.Eventf(wObj, "Warning", "NodeClaimCreationFailed",
-						"Failed to create NodeClaim %s for workspace %s: %v", nodeClaim.Name, wObj.Name, err)
-				} else {
-					c.recorder.Eventf(wObj, "Warning", "NodeClaimCreationFailed",
-						"Failed to create NodeClaim for workspace %s after retries", wObj.Name)
-				}
-				continue // should not return here or expectations will leak
+	if updateErr := workspace.UpdateStatusConditionIfNotMatch(ctx, c.Client, wObj, kaitov1beta1.ConditionTypeNodeClaimStatus, metav1.ConditionFalse,
+		"ScalingUpNodeClaims", fmt.Sprintf("Scaling up %d additional NodeClaims", nodesToCreate)); updateErr != nil {
+		klog.ErrorS(updateErr, "failed to update NodeClaim status condition", "workspace", workspaceKey)
+		return false, fmt.Errorf("failed to update NodeClaim status condition: %w", updateErr)
+	}
+
+	c.expectations.ExpectCreations(c.logger, workspaceKey, nodesToCreate)
+
+	nodeOSDiskSize := c.determineNodeOSDiskSize(wObj)
+
+	for range nodesToCreate {
+		var nodeClaim *karpenterv1.NodeClaim
+		var err error
+		created := false
+
+		retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			nodeClaim = nodeclaim.GenerateNodeClaimManifest(nodeOSDiskSize, wObj)
+			err = c.Client.Create(ctx, nodeClaim)
+
+			if err == nil {
+				created = true
+				return nil
 			}
+			return err
+		})
 
-			klog.InfoS("NodeClaim created successfully",
-				"nodeClaim", nodeClaim.Name,
-				"workspace", klog.KObj(wObj))
+		if !created {
+			// Failed to create, decrement expectations
+			c.expectations.CreationObserved(c.logger, workspaceKey)
+			if err != nil {
```

Comment:
is `err` always != nil if `created` is false?

Reply:
fixed